### ISSUE TITLE
library does not connect to docker service in macbook sylicon

### ISSUE THREAD

**Issue by jesinity** [docker/docker-py#3271]
I tried to use this library without success:

Environment info:

- macos with sylicon M1
- python 3.12
- docker-py 7.10

```
docker version
````

produces:

```
Client:
 Cloud integration: v1.0.35+desktop.13
 Version:           26.1.1
 API version:       1.45
 Go version:        go1.21.9
 Git commit:        4cf5afa
 Built:             Tue Apr 30 11:44:56 2024
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.30.0 (149282)
 Engine:
  Version:          26.1.1
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.9
  Git commit:       ac2de55
  Built:            Tue Apr 30 11:48:04 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.31
  GitCommit:        e377cd56a71523140ca6ae87e30244719194a521
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

when I do:

```
import docker
client = docker.from_env()
```

I get:
 
```

>>> import docker
>>> client = docker.from_env()

Traceback (most recent call last):
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/urllib3/connectionpool.py", line 793, in urlopen
    response = self._make_request(
               ^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/urllib3/connectionpool.py", line 496, in _make_request
    conn.request(
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/urllib3/connection.py", line 400, in request
    self.endheaders()
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/http/client.py", line 1314, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/http/client.py", line 1074, in _send_output
    self.send(msg)
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/http/client.py", line 1018, in send
    self.connect()
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/docker/transport/unixconn.py", line 26, in connect
    sock.connect(self.unix_socket)
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/requests/adapters.py", line 486, in send
    resp = conn.urlopen(
           ^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/urllib3/connectionpool.py", line 847, in urlopen
    retries = retries.increment(
              ^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/urllib3/util/retry.py", line 470, in increment
    raise reraise(type(error), error, _stacktrace)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/urllib3/util/util.py", line 38, in reraise
    raise value.with_traceback(tb)
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/urllib3/connectionpool.py", line 793, in urlopen
    response = self._make_request(
               ^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/urllib3/connectionpool.py", line 496, in _make_request
    conn.request(
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/urllib3/connection.py", line 400, in request
    self.endheaders()
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/http/client.py", line 1314, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/http/client.py", line 1074, in _send_output
    self.send(msg)
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/http/client.py", line 1018, in send
    self.connect()
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/docker/transport/unixconn.py", line 26, in connect
    sock.connect(self.unix_socket)
urllib3.exceptions.ProtocolError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/docker/api/client.py", line 223, in _retrieve_server_version
    return self.version(api_version=False)["ApiVersion"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/docker/api/daemon.py", line 181, in version
    return self._result(self._get(url), json=True)
                        ^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/docker/utils/decorators.py", line 44, in inner
    return f(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/docker/api/client.py", line 246, in _get
    return self.get(url, **self._set_request_timeout(kwargs))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/requests/sessions.py", line 602, in get
    return self.request("GET", url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/requests/adapters.py", line 501, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/docker/client.py", line 94, in from_env
    return cls(
           ^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/docker/client.py", line 45, in __init__
    self.api = APIClient(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/docker/api/client.py", line 207, in __init__
    self._version = self._retrieve_server_version()
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Administrator/miniconda3arm/envs/pydev/lib/python3.12/site-packages/docker/api/client.py", line 230, in _retrieve_server_version
    raise DockerException(
docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
```

**Comment by Khushiyant** [comment]
I tried to reproduce the error without success with above mentioned requirements

**Comment by elliotsegler** [comment]
I don't think this is so much an with docker-py. From docker-desktop for osx 4.13 onwards the behavior for the creation of the `/var/run/docker.sock` socket changed. 

See the following for the gory details:
 - https://github.com/docker/for-mac/issues/6529
 - https://docs.docker.com/desktop/mac/permission-requirements/

TL;DR fix for you is to ensure that under `Settings > Advanced > Allow the default Docker socket to be used (requires password)` is checked.

It would though, be nice if docker-py checked to see if `~/.docker/run/docker.sock` existed for darwin based systems.

**Comment by jesinity** [comment]
that indeed was the issue, I checked that options and everything started working!
